CHANGELOG.md

```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 ### Changed
+- [PMM-4168](https://jira.percona.com/browse/PMM-4168): Updated all dependencies including `github.com/prometheus/golang_client`. 
+As a result a lot of metrics was changed from Counter to Gauge. For more information look at commit [af19190](https://github.com/percona/mongodb_exporter/commit/af1919030e6b9cf3a3cc6da069435f09d49f5099). 
```

Comment:
That commit does not really explain why you changed metric types.
I assume that this is because of Prometheus client removing `Set` method for counters in 0.9.0. But many (most? all?) metrics that were changed to gauges should remain counters for Prometheus to be able to detect resets and for our dashboards to continue working.
I mentioned the proper fix in the other ticket: `NewConstMetric`. See here: https://godoc.org/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics

Reply:
> I assume that this is because of Prometheus client removing Set method for counters in 0.9.0.

Yes.

> But many (most? all?) metrics that were changed to gauges should remain counters 

So many, most, or all? How you propose to increment these values without `Set()` method? Should we save previous value after each collect and then subtract it from the next value, or what?

> to be able to detect resets

Could you explain please what did you mean? How we are detecting resets and for what reasons we are detecting it? How does it work?

> and for our dashboards to continue working

Why they will break? Because of metric types changes? Why we can't update our dashboards in this case? 

> I mentioned the proper fix in the other ticket: NewConstMetric.

`NewConstMetric returns a metric with one fixed value that cannot be changed.`
I'm not sure how it can help. Can you give a little bit more information and context, how it can help and with what exactly?

Reply:
Did you read the link?..

> The Collector example illustrates the use case. You can also look at the source code of the processCollector (mirroring process metrics), the goCollector (mirroring Go metrics), or the expvarCollector (mirroring expvar metrics) as examples that are used in this package itself.

That's what we should do there.

Re counters vs gauges:
* https://prometheus.io/docs/concepts/metric_types/
* https://www.robustperception.io/how-does-a-prometheus-counter-work
* https://prometheus.io/docs/prometheus/latest/querying/functions/#rate

When MongoDB restarts, it resets all metrics. PromQL `rate()` function can detect that for a counter to do the right thing, but not for a gauge. No, we can't fix dashboards if incorrect data is already in.

> How you propose to increment these values without Set() method?

By using `NewConstMetric`. Did you really type that question before finishing reading? Why?

Reply:
> Did you read the link?.

Yes, I read the link but doesn't see examples at this moment.

> By using NewConstMetric. Did you really type that question before finishing reading? Why?

So I should revert all Metrics I have changed and replace all of them to `NewConstMetric()` with type Counter? Am I understand right?

Reply:
> Yes, I read the link but doesn't see examples at this moment.

* https://github.com/prometheus/client_golang/blob/master/prometheus/process_collector.go
* https://github.com/prometheus/client_golang/blob/master/prometheus/go_collector.go
* https://github.com/prometheus/client_golang/blob/master/prometheus/expvar_collector.go

You can also search at PMM code and exporters for `NewConstMetric` examples.

> So I should revert all Metrics I have changed and replace all of them to NewConstMetric() with type Counter? Am I understand right?

Most of them. Some should be gauges (but still with `NewConstMetric`) like time: https://www.robustperception.io/are-increasing-timestamps-counters-or-gauges That's why I said "most". Use your own judgment first, then we can go over corner cases.

Reply:
Ack

Reply:
There is also an example in mongodb_exporter itself: https://github.com/percona/mongodb_exporter/commit/27da9c342847f38cf13276f129b9578b6c6c5e9d